sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -656,3 +705,301 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] with AliasHelpe
+          // EqualNullSafe is used to make sure null key can be joined together. Note
+          // outer referenced attributes can be changed during the outer query optimization.
+          // The equality conditions will also serve as an attribute mapping between new
+          // outer references and domain attributes when rewriting the domain joins.
+          // E.g. if the attribute a is changed to a1, the join condition a' <=> outer(a)
+          // will become a' <=> a1, and we can construct the aliases based on the condition:
+          // DomainJoin [a']        Join Inner
+          // +- InnerQuery     =>   :- InnerQuery
+          //                        +- Aggregate [a1] [a1 AS a']
+          //                           +- OuterQuery
+          val conditions = outerReferenceMap.map {
+            case (o, a) => EqualNullSafe(a, OuterReference(o))
+          }
+          (domainJoin, conditions.toSeq, outerReferenceMap)
+        }
+      } else {
+        // Collect outer references from the current node.
+        val outerReferences = collectOuterReferences(plan.expressions)
+        plan match {
+          case Filter(condition, child) =>
+            val (correlated, uncorrelated) =
+              splitConjunctivePredicates(condition)
+              .partition(containsOuter)
+            val (equality, nonEquality) = correlated.partition(isEquality)
+            // Find equivalent outer reference relations and remove equivalent attributes from
+            // parentOuterReferences since they can be replaced directly by expressions
+            // inside the inner plan.
+            val equivalences = collectEquivalentOuterReferences(equality)
+            // When the results are aggregated, outer references inside the non-equality
+            // predicates cannot be used directly as join conditions with the outer query.
+            val outerReferences = if (aggregated) {
+              collectOuterReferences(nonEquality)
+            } else {
+              AttributeSet.empty
+            }
+            val newOuterReferences = parentOuterReferences ++ outerReferences -- equivalences.keySet
+            val (newChild, joinCond, outerReferenceMap) =
+              decorrelate(child, newOuterReferences, aggregated)
+            // Add the mapping from the current node.
+            val newOuterReferenceMap = outerReferenceMap ++ equivalences
+            // Replace all outer references in non-equality filter conditions using the domain
+            // attributes produced for inner query with aggregates. This step is necessary
+            // for pushing down the non-equality filters into the domain join as join conditions.
+            val (newFilterCond, newJoinCond) = if (aggregated) {
+              val nonEqualityCond = replaceOuterReferences(nonEquality, newOuterReferenceMap)
+              (nonEqualityCond ++ uncorrelated, equality)
+            } else {
+              (uncorrelated, correlated)
```

Comment:
if `aggregated=false`, does it mean the `Filter` can still contain `OuterReference` because `nonEquality` contains `OuterReference`?